openapi-codegen/core/src/main/scala/sttp/tapir/codegen/JsonSerdeGenerator.scala

```diff
@@ -0,0 +1,255 @@
+      .filter(x => !allSchemas.contains(x))
+      .map { s =>
+        val name = s.replace("[", "_").replace("]", "_").replace(".", "_") + "JsonCodec"
+        s"""implicit lazy val $name: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec[$s] =
+           |  com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker.make[$s]""".stripMargin
+      }
+      .mkString("", "\n", "\n")
+
+    // Permits usage of Option/Seq wrapped classes at top level without having to be explicit
+    val jsonSerdeHelpers =
+      s"""
+           |implicit def seqCodec[T: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec]: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec[List[T]] =
+           |  com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker.make[List[T]]
+           |implicit def optionCodec[T: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec]: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec[Option[T]] =
+           |  com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker.make[Option[T]]
+           |""".stripMargin
+    doc.components
+      .map(_.schemas.flatMap {
+        // For standard objects, only generate the schema if it's a 'top level' json schema
+        case (name, _: OpenapiSchemaObject) if jsonParamRefs.contains(name) =>
+          Some(genJsoniterClassSerde(adtInheritanceMap)(name))
+        // For named maps, only generate the schema if it's a 'top level' json schema
+        case (name, _: OpenapiSchemaMap) if jsonParamRefs.contains(name) =>
+          Some(genJsoniterNamedSerde(name))
+        // For enums, generate the serde if it's referenced in any json model
+        case (name, _: OpenapiSchemaEnum) if allTransitiveJsonParamRefs.contains(name) =>
+          Some(genJsoniterEnumSerde(name))
+        // For ADTs, generate the serde if it's referenced in any json model
+        case (name, schema: OpenapiSchemaOneOf) if allTransitiveJsonParamRefs.contains(name) =>
+          Some(generateJsoniterAdtSerde(name, schema, fullModelPath))
+        case (_, _: OpenapiSchemaObject | _: OpenapiSchemaMap | _: OpenapiSchemaEnum | _: OpenapiSchemaOneOf) => None
+        case (n, x) => throw new NotImplementedError(s"Only objects, enums, maps and oneOf supported! (for $n found ${x})")
+      })
+      .map(jsonSerdeHelpers + additionalExplicitSerdes + _.mkString("\n"))
+  }
+
+  private def genJsoniterClassSerde(adtInheritanceMap: Map[String, Seq[String]])(name: String): String = {
+    val uncapitalisedName = name.head.toLower +: name.tail
+    if (adtInheritanceMap.getOrElse(name, Nil).isEmpty)
+      s"""implicit lazy val ${uncapitalisedName}JsonCodec: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec[${name}] = com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker.make($jsoniterBaseConfig)"""
+    else throw new NotImplementedError(s"A class cannot be used both in a oneOf at the top level when using jsoniter serdes at $name")
```

Comment:
So the problem here is that if there's a serde declared for a subtype of the ADT, then the discriminator is not added on serialization. I don't know how to get around this.